Chapter12/Exercise08.Tests/MainWindowTests.cs

```diff
@@ -0,0 +1,189 @@
+            AssertErrorMessageWithIllegalSides("3", "10", "4");
+        }
+
+        [MonitoredTest("Sides that form a triangle should produce a correct area"), Order(6)]
+        public void _6_SidesThatFormTriangleShouldProduceCorrectArea()
+        {
+            AssertHasTextBoxes();
+            Assert.That(_calculateButton, Is.Not.Null); // is provided in template, should be ok
+
+            AssertAreaWithLegalSides(6, "3", "4", "5");
+            AssertAreaWithLegalSides(44.039, "9", "10", "12");
+        }
+
+        [MonitoredTest("Mixing illegal sides and legal sides should clear TextBlocks correctly"), Order(7)]
+        public void _7_MixingIllegalSidesWithLegalSidesShouldClearTextBlocksCorrectly()
+        {
+            AssertHasTextBoxes();
+            Assert.That(_calculateButton, Is.Not.Null); // is provided in template, should be ok
+
+            AssertAreaWithLegalSides(6, "3", "4", "5");
+            AssertErrorMessageWithIllegalSides("3", "10", "4");
+            AssertAreaWithLegalSides(44.039, "9", "10", "12");
+        }
+
+        private void AssertAreaWithLegalSides(double expected, string sideA, string sideB, string sideC)
+        {
+            _sideTextBoxes[0].Text = sideA;
+            _sideTextBoxes[1].Text = sideB;
+            _sideTextBoxes[2].Text = sideC;
+
+            _calculateButton.FireClickEvent();
+
+            Assert.That(_errorTextBlock.Text, Is.EqualTo(string.Empty),
+                () => "Sides that form a triangle should not give an error message");
+
+            double area = double.Parse(_areaTextBlock.Text);
+            Assert.That(area, Is.EqualTo(expected),
+                () => "Sides that form a triangle should produce an area rounded to 3 decimal digits");
+        }
+
+        private void AssertErrorMessageWithIllegalSides(string sideA, string sideB, string sideC)
+        {
+            _sideTextBoxes[0].Text = sideA;
+            _sideTextBoxes[1].Text = sideB;
+            _sideTextBoxes[2].Text = sideC;
+
+            _calculateButton.FireClickEvent();
+
+            Assert.That(_errorTextBlock.Text, Is.Not.EqualTo(string.Empty),
+                () => "Sides that don't form a triangle should give an error message");
+
+            Assert.That(_areaTextBlock.Text, Is.EqualTo(string.Empty),
+                () => "Sides that don't form a triangle should not produce an area");
+        }
+
+        private void AssertHasTextBoxes()
+        {
+            Assert.That(_sideTextBoxes, Is.Not.Null);
+            Assert.That(_sideTextBoxes.Length, Is.EqualTo(3),
+                () => "There should be 3 TextBoxes for entering sides.");
+        }
+
+        private void AssertMaxInvoke(int expected, int a, int b, int c)
+        {
+            object[] parameters = new object[] { a, b, c };
+            object result = _maxFunction.Invoke(_mainWindow.Window, parameters);
+            Assert.That(result, Is.Not.Null);
```

Comment:
I think it would be usefull to add a student friendly error message for these asserts